runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -2937,19 +2937,259 @@ J9::Z::TreeEvaluator::anewArrayEvaluator(TR::Node * node, TR::CodeGenerator * cg
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfContiguousArraySizeField(), cg), 0);
+   generateSILInstruction(cg, use64BitClasses ? TR::InstOpCode::MVGHI : TR::InstOpCode::MVHI, node, generateS390MemoryReference(targetReg, fej9->getOffsetOfDiscontiguousArraySizeField(), cg), 0);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
+
+   // We end up in this region of the ICF if the first dimension is non-zero and the second dimension is zero.
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, nonZeroFirstDimLabel);
+
+   TR::Register *componentClassReg = cg->allocateRegister();
+   generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, componentClassReg, generateS390MemoryReference(classReg, offsetof(J9ArrayClass, componentType), cg));
+
+   // Calculate maximum allowable object size in elements and jump to OOL if it's higher than firstDimLenReg.
+   int32_t elementSize = TR::Compiler->om.sizeofReferenceField();
+   uintptr_t maxObjectSize = cg->getMaxObjectSizeGuaranteedNotToOverflow();
+   uintptr_t maxObjectSizeInElements = maxObjectSize / elementSize;
+
+   if (comp->target().is64Bit() && !(maxObjectSizeInElements >= 0 && maxObjectSizeInElements <= (uintptr_t)INT_MAX))
+      {
+      generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CLG, node, firstDimLenReg, static_cast<int64_t>(maxObjectSizeInElements), TR::InstOpCode::COND_BRH, oolJumpLabel, false);
+      }
+   else
+      {
+      generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::CL, node, firstDimLenReg, static_cast<int32_t>(maxObjectSizeInElements), TR::InstOpCode::COND_BHR, oolJumpLabel, false);
+      }
+
+   // Now check to see if we have enough space to do the allocation. If not then jump to OOL code.
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, temp1Reg, firstDimLenReg);
+
+   int32_t round = (elementSize >= TR::Compiler->om.objectAlignmentInBytes()) ? 0 : TR::Compiler->om.objectAlignmentInBytes();
+   int32_t disp32 = round ? (round - 1) : 0;
+
+   static const uint8_t multiplierToStrideMap[] = {0, 0, 1, 0, 2, 0, 0, 0, 3};
+   generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, temp1Reg, temp1Reg, multiplierToStrideMap[elementSize]);
+   generateRILInstruction(cg, TR::InstOpCode::AGFI, node, temp1Reg, static_cast<int32_t>(TR::Compiler->om.contiguousArrayHeaderSizeInBytes()) + disp32);
+
+   if (round)
+      {
+      generateRILInstruction(cg, TR::InstOpCode::NILF, node, temp1Reg, -round);
+      }
 
+   // temp2Reg = firstDimLenReg * 16 (discontiguousArrayHeaderSizeInBytes)
+   TR::Register *temp2Reg = cg->allocateRegister();
+   generateRRInstruction(cg, TR::InstOpCode::LR, node, temp2Reg, firstDimLenReg);
+   generateRSInstruction(cg, TR::InstOpCode::getShiftLeftLogicalSingleOpCode(), node, temp2Reg, temp2Reg, 4);
```

Comment:
We can fold `firstDimLenReg` into the shift. Also let's not use a magic constant here. As Igor's recent work has shown the value of `discontiguousArrayHeaderSizeInBytes` can change in the future, so we don't want to hunt this bug down. Use the API.

Reply:
I've folded the register into the shift in instruction in 0c0a523.

We also use the API now as you suggested. I've added in an assert to make sure that the size returned is 16 (see 0c0a523). If it changes in the future then we will catch this immediately. Alternatively we can do the calculation here to figure out the correct shift amount no matter what size is returned if you think doing that extra work is worthwhile.